### PR TITLE
`azurerm_storage_account_customer_managed_key` - support for cross-tenant customer-managed keys

### PR DESCRIPTION
**Pull request by Sewci0**
Adds support for Cross Tenant CMK 

https://learn.microsoft.com/en-us/azure/storage/common/customer-managed-keys-configure-cross-tenant-new-account

Introduces two new properties:
* `key_vault_uri` - used when the SP has no access to the vault that stores the CMK 
* `federated_identity_client_id` - points at the appID used for federated access

### CODE REVIEW

File: website/docs/r/storage_account_customer_managed_key.html.markdown
Comment (magodo) [inline website/docs/r/storage_account_customer_managed_key.html.markdown:103]: Shall we move this line to the description of `federated_identity_client_id`?

File: internal/services/storage/storage_account_customer_managed_key_resource.go
Comment (magodo) [inline internal/services/storage/storage_account_customer_managed_key_resource.go:62]: Assuming with the right role being assigned to the cross tenant app (from the provisioner), the provisioner should has the permission to read the keyvaults and get its URI by its ID, so we can eliminate this new property?

File: internal/services/storage/storage_account_customer_managed_key_resource.go
Comment (magodo) [inline internal/services/storage/storage_account_customer_managed_key_resource.go:193]: Probably we shall optioanlly set this only when `federated_identity_client_id` is specified in the config?

**Review by magodo — CHANGES_REQUESTED** [review]
Thanks for this PR!
I've taken a look through and left some comments inline. Meanwhile, there needs one/more acctests to cover this new change.

File: internal/services/storage/storage_account_customer_managed_key_resource.go
Comment (Sewci0) [reply internal/services/storage/storage_account_customer_managed_key_resource.go:62]: I don't believe caller has any access to that Key Vault. The ad app is using a separate federated identity to access it. Using `id` would required sharing additional informations between tenants i.e. `subscription ID` and `resource group name` which isn't necessary for the setup.

This somehow mirrors Azure's UI where you can either select the key vault by id or using the URI.

**Comment by Sewci0** [comment]
Thank you @magodo for looking into it, I've replied to your comment and fixed the rest of the issues. 

I've also added acceptance tests but these are super tricky to run as you need to set up two separate tenants.

**Comment by magodo** [comment]
@Sewci0 Thank you for the update! It now LGTM!

**Review by katbyte — CHANGES_REQUESTED** [review]
We have some test failures:
```
------- Stdout: -------
=== RUN   TestAccStorageAccountCustomerManagedKey_updateKey
=== PAUSE TestAccStorageAccountCustomerManagedKey_updateKey
=== CONT  TestAccStorageAccountCustomerManagedKey_updateKey
    testcase.go:110: Step 1/4 error: After applying this test step, the plan was not empty.
        stdout:
        
        
        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
          ~ update in-place
        
        Terraform will perform the following actions:
        
          # azurerm_storage_account_customer_managed_key.test will be updated in-place
          ~ resource "azurerm_storage_account_customer_managed_key" "test" {
                id                 = "/subscriptions/*******/resourceGroups/acctestRG-230224173716578224/providers/Microsoft.Storage/storageAccounts/acctestsai40tk"
              - key_vault_uri      = "https://acctestkvi40tk.vault.azure.net/" -> null
                # (4 unchanged attributes hidden)
            }
        
        Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccStorageAccountCustomerManagedKey_updateKey (380.72s)
FAIL
```

**Comment by vaishakhbn** [comment]
Hi, do you know if this PR is being actively worked on?

**Comment by Sewci0** [comment]
Hi @magodo, apologies it has taken so long but I've fixed the tests and rerun them locally. It should be ready to go. Thanks!

File: internal/services/storage/storage_account_customer_managed_key_resource.go
Comment (tombuildsstuff) [inline internal/services/storage/storage_account_customer_managed_key_resource.go:62]: FWIW we're intentionally not exposing `key_vault_uri` as a field across the Provider, since we're unable to determine within the Provider when the Key Vault is (temporarily) inaccessible or (permanently) gone - so unfortunately that's not something we can ship without introducing other issues (and even if we made an exception for this resource, we'd get asks for this on other resources, which would have the same issue).

Instead, we should be able to look this up given the `tenant_id`, which would solve this in a different manner - WDYT?

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @Sewci0 

Thanks for rebasing this / apologies for the delayed review here.

Taking a look through here unfortunately we're unable to expose a schema field for the Key Vault URI - we'd need to look this up via Resource Manager to be able to ascertain when the Key Vault exists / is temporarily gone, so is it possible to look this up using the Tenant ID instead?

Thanks!

**Comment by Sewci0** [comment]
@tombuildsstuff I've fixed the formatting errors. Let me know your thoughts on my comment in https://github.com/hashicorp/terraform-provider-azurerm/pull/20356#pullrequestreview-1418790332 Thanks!

**Comment by Sewci0** [comment]
@katbyte Sorry forgot to tag you after fixing the tests. Acceptance tests are now all passing on my side. Would you mind taking a look? Thank you!

**Comment by Sewci0** [comment]
Hi @tombuildsstuff, have you had a chance to look into this?

**Comment by vaishakhbn** [comment]
Hi @tombuildsstuff , @Sewci0, we are eagerly awaiting the release of cross-tenant CMK using terraform provider :)

File: website/docs/r/storage_account_customer_managed_key.html.markdown
Comment (manicminer) [inline website/docs/r/storage_account_customer_managed_key.html.markdown:128]:
```suggestion
* `key_vault_id` - (Optional) The ID of the Key Vault. Exactly one of `key_vault_id`, or `key_vault_uri` must be specified.

~> Note: When the principal running Terraform has access to the subscription containing the Key Vault, it's recommended to use the `key_vault_id` property for maximum compatibility, rather than the `key_vault_uri` property.

```

File: internal/services/storage/storage_account_customer_managed_key_resource.go
Comment (manicminer) [reply internal/services/storage/storage_account_customer_managed_key_resource.go:62]: Thanks for the explanation @Sewci0. Looking at the [docs](https://learn.microsoft.com/en-us/azure/storage/common/customer-managed-keys-configure-cross-tenant-new-account?tabs=azure-portal#create-a-new-storage-account-encrypted-with-a-key-from-a-different-tenant) for this configuration, it seems like referencing the URI is the best/only way forward in this case, so I think it's ok to add this.

**Review by manicminer — APPROVED** [review]
@Sewci0 Thanks for keeping this PR updated. The implementation looks good to me, but we'll need to set up a few things in order to be able to test this change in CI going forward.

In the meantime, I've approved the PR but before merging I'll work on setting up the test infrastructure. I'll only get to this in the next week or so as I'm away after today, but we should be able to get this merged pretty soon.

In view of the concerns highlighted earlier, I've suggested a note in the docs recommending that users prefer the `key_vault_id` property where it's feasible to do so.

**Comment by Sewci0** [comment]
Hi @manicminer, I've rebased the branch once again and fixed some conflicts. Let me know how the testing's going. Thx!

**Comment by msmygit** [comment]
@tombuildsstuff @magodo @manicminer -- any idea if this PR is being looked at activtely and will be incorporated? This has been pending for quite a long time now with more and more people awaiting its release. Thank you in advance!

**Review by manicminer — APPROVED** [review]
@Sewci0 Thank you for keeping the branch updated and for your patience whilst we update our testing infrastructure. I've pushed a few minor changes to the test config and it LGTM!

Since I've updated our TeamCity configuration, I'll let another maintainer look over those changes before merging.

Screenshot 2023-09-21 at 00 43 16

**Comment by manicminer** [comment]
TeamCity config changes for additional review: https://github.com/hashicorp/terraform-provider-azurerm/pull/20356/commits/707a94b703d6eb0e604235211feae434ca10b1a6

**Review by stephybun — APPROVED** [review]
Tests look good. Thanks for this @Sewci0 and @manicminer! LGTM 🎉

**Comment by Sewci0** [comment]
@manicminer This is very exciting. Thank you!

**Pull request merged**
